src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,411 @@
+    private long solvingStart;
+    private long solvingEnd;
+
+    @Override
+    public InferenceSolution solve(Map<String, String> configuration, Collection<Slot> slots,
+            Collection<Constraint> constraints, QualifierHierarchy qualHierarchy,
+            ProcessingEnvironment processingEnvironment) {
+
+        InferenceSolution solution = null;
+
+        configure(configuration);
+        configureLattice(qualHierarchy);
+        Serializer<?, ?> defaultSerializer = createSerializer(backEndType, lattice);
+
+        if (useGraph) {
+            this.graphBuildingStart = System.currentTimeMillis();
+            this.constraintGraph = generateGraph(slots, constraints, processingEnvironment);
+            this.graphBuildingEnd = System.currentTimeMillis();
+            StatisticRecorder.record(StatisticKey.GRAPH_GENERATION_TIME, (graphBuildingEnd - graphBuildingStart));
+            solution = graphSolve(constraintGraph, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, defaultSerializer);
+        } else {
+            realBackEnd = createBackEnd(backEndType, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, lattice, defaultSerializer);
+            solution = solve();
+        }
+
+        if (solution == null) {
+            ErrorReporter.errorAbort("Null solution detected!");
+        }
+
+        if (collectStatistic) {
+            Map<String, Integer> modelRecord = recordSlotConstraintSize(slots, constraints);
+            PrintUtils.printStatistic(StatisticRecorder.getStatistic(), modelRecord, backEndType,
+                    useGraph, solveInParallel);
+            PrintUtils.writeStatistic(StatisticRecorder.getStatistic(), modelRecord, backEndType,
+                    useGraph, solveInParallel);
+        }
+        return solution;
+    }
+
+    /**
+     * This method configures following arguments: backEndType, useGraph,
+     * solveInParallel, and collectStatistic
+     * 
+     * @param configuration
+     */
+    private void configure(final Map<String, String> configuration) {
+
+        final String backEndName = configuration.get(Constants.BACK_END_TYPE);
+        final String useGraph = configuration.get(Constants.USE_GRAPH);
+        final String solveInParallel = configuration.get(Constants.SOLVE_IN_PARALLEL);
+        final String collectStatistic = configuration.get(Constants.COLLECT_STATISTIC);
+
+        if (backEndName == null || backEndName.equals("MaxSAT")) {
+            // Configure back end type. Default is MaxSAT.
+            this.backEndType = Constants.MAX_SAT;
+        } else if (backEndName.equals("Lingeling")) {
+            this.backEndType = Constants.LINGELING;
+        } else if (backEndName.equals("LogiQL")) {
+            this.backEndType = Constants.LOGIQL;
+        } else {
+            ErrorReporter.errorAbort("Back end \"" + backEndName + "\" has not been implemented yet.");
```

Comment:
> Why do you want to extend this class when a new back end is added?
My use case is when developing a prototype backend for Ontology, I try to avoid touch the GeneralSolver class, but currently I cannot do this. In general, it is a question of if should us support user to have one's own backend plugged into the solver framework, if doesn't, then I agree with using predefined set to control backend type.

In addition, it would be better to decoupling the GeneralSolver class with specific backEndTypes in both cases, e.g. eliminating code like `if (backEndName.equals("aSpecificBKType")`. Instead, having code more general like using `(BackEndType.values().contains(backEndName))` might be better (BackEndType is an enum class for controlling BackEndTypes).